src/ledger/ledger.js

```diff
@@ -0,0 +1,344 @@
+
+/**
+ * The ledger for storing identity changes and (eventually) Grain distributions.
+ *
+ * The following API methods change the ledger state:
+ * - `addUser`
+ * - `renameUser`
+ * - `removeUser`
+ * - `addAlias`
+ * - `removeAlias`
+ *
+ * Every time the ledger state is changed, a corresponding Action is added to
+ * the ledger's action log. The ledger state may be serialized by saving the
+ * action log, and then reconstructed by replaying the action log. The
+ * corresponding methods are `actionLog` and `Ledger.fromActionLog`.
+ *
+ * It's important that any API method that fails (e.g. trying to add a
+ * conflicting user) fails without mutating the ledger state; this way we avoid
+ * ever getting the ledger in a corrupted state. To make this easier to test,
+ * the test code uses deep equality testing on the ledger before/after
+ * attempting illegal actions. To ensure that this testing works, we should
+ * avoid adding any ledger state that can't be verified by deep equality
+ * checking (e.g. don't store state in functions or closures that aren't
+ * attached to the Ledger object).
+ *
+ * Currently, the Ledger only supports user management; logic for tracking
+ * Grain distributions will be added in a future commit.
+ */
+export class Ledger {
+  _actionLog: Action[];
+  _users: Map<UserId, User>;
+  _usernameToId: Map<Username, UserId>;
+  _aliases: Set<NodeAddressT>;
+
+  constructor() {
+    this._actionLog = [];
+    this._users = new Map();
+    this._usernameToId = new Map();
+    this._aliases = new Set();
+  }
+
+  /**
+   * Return all of the Users in the Ledger.
+   */
+  users(): $ReadOnlyArray<User> {
+    return Array.from(this._users.values());
+  }
+
+  /**
+   * Return the User matching given id, or undefined if no user
+   * matches the id.
+   */
+  userById(id: UserId): ?User {
+    return this._users.get(id);
+  }
+
+  /**
+   * Return the User matching given username, or undefined
+   * if no user matches the username.
+   */
+  userByUsername(name: Username): ?User {
+    const id = this._usernameToId.get(name);
+    if (id != null) {
+      return this._users.get(id);
+    }
+  }
+
+  /**
+   * Add a User to the Ledger.
+   *
+   * This will reserve the user's username, and its innate address.
+   *
+   * Will no-op is the user is already in the ledger.
+   *
+   * Will fail if a conflicting user is present, or if the user's username is
+   * already taken, or if the user's innate address is already taken, or if any
+   * of the user's aliases are already taken.
+   */
+  addUser(user: User): Ledger {
+    return this._act({type: "ADD_USER", user: user});
+  }
+  _addUser(user: User) {
+    const existingUser = this._users.get(user.id);
+    if (existingUser != null) {
+      if (deepEqual(existingUser, user)) {
+        // No-op; This user already exists in the ledger.
+        return;
+      } else {
+        throw new Error(`addUser: conflicting user with id ${user.id}`);
+      }
+    }
```

Comment:
This seems fairly awkward to me and an example of why the 1:1 relation between API and events doesn't improve things. The UserId is it's unique reference, the other fields are state. So this is like a partial idempotence? Where a User in the same state as the existing one is silently ignored, but the same semantical user in a divergent state fails because this method wasn't supposed to check for differences while it could.

Abandoning the 1:1 mapping, this could actually be made much simpler and idempotent. In pseudo:

```js
trackUser(user: User): $ReadOnlyArray<Event> {
  // if user does not exist
    // return UserCreated event

  // const changes = [];

  // if username is different
    // add UsernameChanged to changes

  // if aliases are different
    // add AliasesChanged to changes

  // return changes
}
```

An existing user in the same state would be ignored (empty array of events). Changes would be reflected by change events. New users would have a create event.

Reply:
Yeah, it's awkward, I changed this so that the client doesn't create the Users.